### ISSUE TITLE
Moving down into an empty object or list fails.

### ISSUE THREAD

**Issue by luke-clifton** [qfpl/waargonaut#38]
This isn't what I was expecting.

```
D.withCursor $ \c -> D.down c >> pure ()
```

will fail on `"{}"` or `"[]"`. I was expecting the `D.down` to succeed, but then any `D.moveRight` to cause an error. Is this documented somewhere?

**Comment by mankyKitty** [comment]
There isn't anything in the index for that position, so it is considered a failure. I don't think I have explicit documentation for this behaviour, I'll add some in.

**Comment by luke-clifton** [comment]
So, what I want to do is, go `down` into an object, then do a `foldCursor` (or a `rightwardSnoc`). I'm perfectly happy for that to just give me the initial value of the accumulator if the object is empty. What is the alternative? If I check for failure on the `down` I still don't know why that is, it could be because it's a text or numeric field.

Or, rephrased, how can I check that my cursor is currently pointed at an object without parsing it all into a `Json`?

**Comment by luke-clifton** [comment]
Which, looking at the code just now, is exactly what your [`listAt`](http://hackage.haskell.org/package/waargonaut-0.2.0.1/docs/src/Waargonaut.Decode.html#listAt) is trying to do. So, if you `listAt` and the cursor is sitting in front of a text or numeric field, it will result in an empty list, which seems slight counter-intuitive.

**Comment by mankyKitty** [comment]
That `listAt` behaviour seems a bit more 'bug' than 'counter-intuitive', so thanks for that !

Currently there isn't a direct way to check if you're currently looking at an object or other specific type, without going to the `Json` structure then using the lenses and prisms to nose around. That's a useful feature though and wouldn't be difficult to include.

To keep you moving, you could take advantage of the `Alt` instance for `Decoder` to combine alternatives for what may exist at a given cursor position. Until I fix the `listAt` this will mean you need to be aware of ordering considerations, which isn't awesome.

Alternatively, you are able to use the functions from the library that powers the cursor itself `hw-json`. Specifically from the module [HaskellWorks.Data.Json.Type](https://hackage.haskell.org/package/hw-json-0.9.0.1/docs/HaskellWorks-Data-Json-Type.html#t:JsonTypeAt).

```haskell
jsonTypeAt (unJCurs cursor) :: Maybe JsonType
```

Using `Alt` would be my first preference though. Does any of that help?

**Comment by luke-clifton** [comment]
It seems that exposing [`jsonTypeAt`](http://hackage.haskell.org/package/hw-json-0.9.0.1/docs/HaskellWorks-Data-Json-Type.html#v:jsonTypeAt) is what I want.

**Comment by luke-clifton** [comment]
Ha, we posted at the same time. :+1:

**Comment by mankyKitty** [comment]
I'll leave this open until I've patched `listAt` others if I find more

**Comment by luke-clifton** [comment]
Fixed by #39
